pkgs/development/python-modules/home-assistant-chip-wheels/default.nix

```diff
@@ -0,0 +1,305 @@
+{
+  appdirs,
+  appnope,
+  black,
+  build,
+  clang-tools,
+  click,
+  colorama,
+  coloredlogs,
+  coverage,
+  cryptography,
+  diskcache,
+  fetchFromGitHub,
+  glib,
+  gn1924,
+  googleapis-common-protos,
+  google-cloud-storage,
+  ipython,
+  jinja2,
+  json5,
+  jsonschema,
+  lark,
+  lib,
+  libnl,
+  mobly,
+  mypy,
+  mypy-extensions,
+  mypy-protobuf,
+  ninja,
+  packaging,
+  parameterized,
+  pip-tools,
+  pkg-config,
+  prompt-toolkit,
+  protobuf,
+  psutil,
+  ptpython,
+  pyelftools,
+  pygments,
+  pykwalify,
+  pylint,
+  pyperclip,
+  pyserial,
+  python3,
+  python-daemon,
+  pyyaml,
+  requests,
+  setuptools,
+  six,
+  sphinx,
+  sphinx-argparse,
+  sphinx-design,
+  stdenv,
+  stringcase,
+  toml,
+  tornado,
+  types-protobuf,
+  types-pyyaml,
+  types-requests,
+  types-setuptools,
+  watchdog,
+  websockets,
+  wheel,
+  yapf,
+  zap-chip,
+}:
+
+stdenv.mkDerivation rec {
+  pname = "home-assistant-chip-wheels";
+  version = "2024.9.0";
+  src = fetchFromGitHub {
```

Comment:
While this approach shows great ingenuity, I think compressing and then decompressing this is fairly undesirable. Earlier this year, I was able to check out the full source tree with ~800MB. I need to do some digging to figure out how I did it, would you mind waiting to merge this for another day to give me time to investigate?

Reply:
It looks like we can get this down to ~487MB:

```
du --apparent-size --human-readable /nix/store/x51kgcdzr8wgkcbk9hqymlmq5vm5km9g-source/ | tail -2
486M    /nix/store/x51kgcdzr8wgkcbk9hqymlmq5vm5km9g-source/connectedhomeip
487M    /nix/store/x51kgcdzr8wgkcbk9hqymlmq5vm5km9g-source/
```

The following patch will only check out needed submodules and copy everything in a shallow manner:

```
diff --git a/pkgs/development/python-modules/home-assistant-chip-wheels/default.nix b/pkgs/development/python-modules/home-assistant-chip-wheels/default.nix
index 5ca168c56de5..df6eec748c02 100644
--- a/pkgs/development/python-modules/home-assistant-chip-wheels/default.nix
+++ b/pkgs/development/python-modules/home-assistant-chip-wheels/default.nix
@@ -11,6 +11,7 @@
   cryptography,
   diskcache,
   fetchFromGitHub,
+  git,
   glib,
   gn1924,
   googleapis-common-protos,
@@ -72,27 +73,25 @@ stdenv.mkDerivation rec {
     owner = "home-assistant-libs";
     repo = "chip-wheels";
     rev = version;
-    fetchSubmodules = true;
-    name = "home-assistant-chip-wheels-${version}.tar.gz";
-    hash = "sha256-iAren1/qEPymsUH2hQ8ZITfkCYwZlravU9BLK90PTpU=";
-
+    fetchSubmodules = false;
+    leaveDotGit = true;
+    hash = "sha256-T0G6mxb/5wFOxPLL92Ay34oP+9Xvk9w0YV9VSzWJuzw=";
     postFetch = ''
-      # remove large folders which are unneeded by the python bindings
-      rm -r $out/connectedhomeip/third_party/openthread/ot*
-      rm -r $out/connectedhomeip/third_party/ti_simplelink_sdk/repo_cc*
-      rm -r $out/connectedhomeip/third_party/silabs/gecko_sdk
+      export PATH="${python3}/bin:${git}/bin:$PATH"
 
-      # compress to not exceed the 2GB output limit
-      # try to make a deterministic tarball
-      tar -I 'gzip -n' \
-      --sort=name \
-      --mtime=1970-01-01 \
-      --owner=0 --group=0 \
-      --numeric-owner --mode=go=rX,u+rw,a-s \
-      --transform='s@^@source/@S' \
-      -cf temp  -C "$out" .
-      rm -r "$out"
-      mv temp "$out"
+      cd $out
+      # Download connectedhomeip.
+      git fetch
+      git reset --hard HEAD
+      git submodule update --init --depth 1 connectedhomeip
+
+      # Initialize only necessary submodules.
+      cd connectedhomeip
+      python3 scripts/checkout_submodules.py --platform linux --shallow
+
+      # Keep the output deterministic.
+      cd $out
+      rm -rf .git/
     '';
   };
```